File: frontend/src/types/verification.ts
/**
 * ARCH-13 verification triage DTOs.
 * Mirrors `app/schemas/verification.py` and `app/models/verification.py`.
 */

export type VerificationStatus =
  | "PENDING"
  | "AGREED"
  | "DISAGREED"
  | "REVIEWED"
  | "AUTO_APPROVED";

export type DisagreementKind = "MISSING" | "CONFLICT" | "FORMAT";

export interface VerificationFieldResponse {
  readonly field_path: string;
  readonly agreed: boolean;
  readonly confidence: string;
  readonly consensus_value: unknown;
  readonly agent_values: readonly unknown[];
  readonly disagreement_kind: DisagreementKind | null;
  readonly resolved_value: unknown;
}

export interface VerificationSummaryResponse {
  readonly id: string;
  readonly work_item_id: string;
  readonly status: VerificationStatus;
  readonly agent_count: number;
  readonly agreement_score: string | null;
  readonly confidence: string | null;
  readonly cost_micros: number;
  readonly auto_approved: boolean;
  readonly reviewed_by_user_id: string | null;
  readonly reviewed_at: string | null;
  readonly created_at: string;
}

export interface VerificationDetailResponse
  extends VerificationSummaryResponse {
  readonly fields: readonly VerificationFieldResponse[];
  readonly details: Record<string, unknown>;
}

export interface VerificationResolveRequest {
  readonly values: Record<string, unknown>;
}

export interface VerificationListParams {
  readonly status?: VerificationStatus;
  readonly skip?: number;
  readonly limit?: number;
}

export const parseScore = (value: string | null | undefined): number | null => {
  if (value === null || value === undefined) {
    return null;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
};

export const formatFieldValue = (value: unknown): string => {
  if (value === null || value === undefined) {
    return "—";
  }
  if (typeof value === "string") {
    return value.length === 0 ? "(empty)" : value;
  }
  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  try {
    return JSON.stringify(value);
  } catch {
    return "(unrenderable)";
  }
};

export const isBlocking = (status: VerificationStatus): boolean =>
  status === "PENDING" || status === "DISAGREED";

File: backend/app/schemas/verification.py
"""ARCH-13 Step 13.8 — verification API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.verification import DisagreementKind, VerificationStatus


class VerificationFieldResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    field_path: str
    agreed: bool
    confidence: Decimal
    consensus_value: Optional[Any] = None
    agent_values: list[Any] = Field(default_factory=list)
    disagreement_kind: Optional[DisagreementKind] = None
    resolved_value: Optional[Any] = None


class VerificationSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    work_item_id: uuid.UUID
    status: VerificationStatus
    agent_count: int
    agreement_score: Optional[Decimal] = None
    confidence: Optional[Decimal] = None
    cost_micros: int
    auto_approved: bool
    reviewed_by_user_id: Optional[uuid.UUID] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime


class VerificationDetailResponse(VerificationSummaryResponse):
    fields: list[VerificationFieldResponse] = Field(default_factory=list)
    details: dict[str, Any] = Field(default_factory=dict)


class VerificationResolveRequest(BaseModel):
    """The reviewer's chosen value for each disagreed field."""

    values: dict[str, Any] = Field(
        ...,
        description=(
            "field_path -> chosen value. Must cover every disagreed field and "
            "no others."
        ),
    )

    @field_validator("values")
    @classmethod
    def non_empty(cls, value: dict[str, Any]) -> dict[str, Any]:
        if not value:
            raise ValueError(
                "values must not be empty. A resolve with no chosen values is "
                "an approval nobody made."
            )
        for key, chosen in value.items():
            if not isinstance(key, str) or not key.strip():
                raise ValueError(f"Invalid field path: {key!r}")
            if isinstance(chosen, (dict, list)):
                raise ValueError(
                    f"Field {key!r} was given a {type(chosen).__name__}. "
                    "Resolved values are scalars; a nested object is a "
                    "document fragment."
                )
        return value


__all__ = [
    "VerificationDetailResponse",
    "VerificationFieldResponse",
    "VerificationResolveRequest",
    "VerificationSummaryResponse",
]
